src/v2/controllers/vreplicaset_controller/trusted/rely_guarantee.rs

```diff
@@ -0,0 +1,164 @@
+                let owner_references = etcd_obj.metadata.owner_references.get_Some_0();
+                &&& s.resources().contains_key(req.key)
+                &&& etcd_obj.metadata.resource_version.is_Some()
+                &&& etcd_obj.metadata.resource_version
+                    == req.preconditions.get_Some_0().resource_version
+                &&& etcd_obj.metadata.owner_references.is_Some()
+                &&& exists |vrs: VReplicaSetView| 
+                    #[trigger] owner_references.contains(vrs.controller_owner_ref())
+            }
+    }
+}
+
+pub open spec fn vrs_rely(other_id: int) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        forall |msg| {
+            &&& #[trigger] s.in_flight().contains(msg)
+            &&& msg.content.is_APIRequest()
+            &&& msg.src == HostId::Controller(other_id)
+        } ==> match msg.content.get_APIRequest_0() {
+            APIRequest::CreateRequest(req) => vrs_rely_create_req(req)(s),
+            APIRequest::UpdateRequest(req) => vrs_rely_update_req(req)(s),
+            APIRequest::UpdateStatusRequest(req) => vrs_rely_update_status_req(req)(s),
+            APIRequest::DeleteRequest(req) => vrs_rely_delete_req(req)(s),
+            _ => true,
+        }
+    }
+}
+
+// VRS Guarantee Condition
+
+// VRS only creates pods owned by a VReplicaSet.
+pub open spec fn vrs_guarantee_create_req(req: CreateRequest) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        let owner_references = req.obj.metadata.owner_references.get_Some_0();
+        &&& req.obj.kind == Kind::PodKind
+        &&& req.obj.metadata.owner_references.is_Some()
+        &&& exists |vrs: VReplicaSetView| 
+            owner_references == seq![#[trigger] vrs.controller_owner_ref()]
+    }
+}
+
+// VRS only sends delete requests to pods that carry a resource version precondition,
+// and if the resource version carried matches the one in the corresponding etcd
+// object, then that object is owned by a VReplicaSet.
+pub open spec fn vrs_guarantee_delete_req(req: DeleteRequest) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        let etcd_obj = s.resources()[req.key];
+        &&& req.key.kind == Kind::PodKind
+        &&& req.preconditions.is_Some()
+        &&& req.preconditions.get_Some_0().resource_version.is_Some()
+        &&& {
+            &&& s.resources().contains_key(req.key)
+            &&& etcd_obj.metadata.resource_version 
+                == req.preconditions.get_Some_0().resource_version
+        } ==> {
+            let controller_owners = etcd_obj.metadata.owner_references
+                .get_Some_0()
+                .filter(|o: OwnerReferenceView| {
+                    o.controller.is_Some() && o.controller.get_Some_0()
+                });
+            &&& etcd_obj.metadata.owner_references.is_Some()
+            &&& exists |vrs: VReplicaSetView| 
+                controller_owners == seq![#[trigger] vrs.controller_owner_ref()]
```

Comment:
This seems too strong? The pod VRS controller deletes might have more than one owners, but only one controller owner (which is VRS)

Reply:
Look again; `controller_owners` is filtered precisely that way.

Reply:
I double checked the implementation of the reference RS controller: it manages pods whose controller owner is RS, even if the pods have other owners. See: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_ref_manager.go#L70-L75

Reply:
Perhaps I should have explained further originally. `controller_owners` is defined precisely as the subset of `owner_references` that are controller owners. The line enforces that the only _controller_ owner is a VReplicaSet. My definition allows other non-controller owners.